File: src/scene_manager.js
const HEADER = [0xF0, 0x43, 0x20, 0x7E];
const SIGNATURE = [0x4C, 0x4D, 0x20, 0x20, 0x38, 0x43, 0x39, 0x33]; // "LM  8C93"
const CMD_BULK_REQUEST = 0x6D;
const FOOTER = 0xF7;

class SceneManager {
    constructor() {
        this.scenes = [];
        this.currentScene = null;
        this.activeSceneIndex = 0;
        this.isSyncing = false;
        this.io = null;
    }


    setIO(ioInstance) {
        this.io = ioInstance;
    }

    buildBulkRequest(type, index) {
        return [...HEADER, ...SIGNATURE, CMD_BULK_REQUEST, type, index, FOOTER];
    }

    fetchScenes(midiEngine) {
        return new Promise((resolve) => {
            if (this.isSyncing) {
                resolve();
                return;
            }
            this.isSyncing = true;
            this.scenes = [];
            this.currentScene = null;

            console.log('\n📚 [Scene Manager] Iniciando sincronização da Biblioteca de Cenas...');
            let queue = [];
        
        // Pede o "Current Edit Buffer" (Index 0, type 0x02)
        queue.push({ type: 0x02, index: 0 });

        // Pede as cenas 1 a 99 (Library, type 0x00)
        for (let i = 1; i <= 99; i++) {
            queue.push({ type: 0x00, index: i });
        }

        let sentCount = 0;
        const intervalId = setInterval(() => {
            if (queue.length === 0) {
                clearInterval(intervalId);
                console.log('✅ [Scene Manager] Todas as requisições de cena enviadas, aguardando finalização dos dumps...');
                
                // Aguarda um pequeno tempo para os últimos dumps chegarem
                setTimeout(() => {
                    this.isSyncing = false;
                    console.log(`✅ [Scene Manager] Sincronização concluída! ${this.scenes.filter(Boolean).length} cenas carregadas.`);
                    if (this.io) {
                        this.io.emit('scenesUpdated', this.getState());
                    }
                    resolve();
                }, 2000);
                return;
            }

            const req = queue.shift();
            midiEngine.send(this.buildBulkRequest(req.type, req.index));
            sentCount++;
            
            if (sentCount % 20 === 0) {
                console.log(`⏳ [Scene Manager] Progresso: ${sentCount}/100...`);
            }
        }, 50); // 50ms para suportar a carga de dumps maiores (>500b)
        });
    }

    handleMIDIData(message) {
        if (!message || message.length <= 20) return false;

        // Header: F0 43 00 7E ... (Type Response=00) ... 6D
        if (message[0] === 0xF0 && message[1] === 0x43 && message[14] === 0x6D) {
            const type = message[15];
            const index = message[16];

            // Respostas Type 00 (Library) ou Type 02 (Edit Buffer)
            if ((type === 0x00 || type === 0x02) && message.length > 21) {
                // Nome comeca no offset 20 e tem tamanho 16
                let name = '';
                for (let i = 0; i < 16; i++) {
                    const charCode = message[20 + i];
                    if (charCode >= 32 && charCode <= 126) {
                        name += String.fromCharCode(charCode);
                    } else if (charCode !== 0) {
                        name += ' ';
                    }
                }
                name = name.trim();

                const sceneData = { index, name };

                if (type === 0x02) {
                    this.currentScene = {
                        index: this.activeSceneIndex || 0, // Usa o index ativamente armazenado
                        name: sceneData.name
                    };
                    if (this.io) this.io.emit('currentScene', this.currentScene);
                } else {
                    this.scenes[index] = sceneData;
                }

                return true; // Mensagem consumida com sucesso
            }
        }
        return false;
    }

    getState() {
        return {
            currentScene: this.currentScene,
            scenes: this.scenes.filter(s => s != null) // Retorna lista compactada
        };
    }

    getScenes() {
        return this.scenes.filter(s => s != null);
    }

    getCurrentScene() {
        if (!this.currentScene) return null;
        return {
            index: this.activeSceneIndex || 0,
            name: this.currentScene.name
        };
    }

    setActiveScene(index) {
        this.activeSceneIndex = index;
        if (this.currentScene) {
            this.currentScene.index = index;
        }
    }
}


module.exports = new SceneManager();

File: server_rust/src/scene_manager.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SceneData {
    pub index: u8,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SceneManagerState {
    #[serde(rename = "currentScene")]
    pub current_scene: Option<SceneData>,
    pub scenes: Vec<SceneData>,
}

#[derive(Debug, Clone, Default)]
pub struct SceneManager {
    pub scenes: Vec<Option<SceneData>>,
    pub current_scene: Option<SceneData>,
    pub active_scene_index: u8,
    pub is_syncing: bool,
}

impl SceneManager {
    pub fn new() -> Self {
        Self {
            scenes: vec![None; 100],
            current_scene: None,
            active_scene_index: 0,
            is_syncing: false,
        }
    }

    pub fn handle_midi_data(&mut self, message: &[u8]) -> bool {
        if message.len() <= 20 {
            return false;
        }

        if message[0] == 0xF0 && message[1] == 0x43 && message[14] == 0x6D {
            // Verifica assinatura "LM  8C93" nos bytes 6 a 13 para evitar falsos positivos
            let signature = [0x4C, 0x4D, 0x20, 0x20, 0x38, 0x43, 0x39, 0x33];
            if message[6..14] != signature {
                return false;
            }

            let req_type = message[15];
            let index = message[16];

            if (req_type == 0x00 || req_type == 0x02) && message.len() > 21 {
                let mut name = String::new();
                for i in 0..16 {
                    if 20 + i < message.len() {
                        let c = message[20 + i];
                        if c >= 32 && c <= 126 {
                            name.push(c as char);
                        } else if c != 0 {
                            name.push(' ');
                        }
                    }
                }
                let name = name.trim().to_uppercase();

                if req_type == 0x02 {
                    let scene_data = SceneData {
                        index: self.active_scene_index,
                        name: name.clone(),
                    };
                    self.current_scene = Some(scene_data);
                } else if (index as usize) < self.scenes.len() {
                    self.scenes[index as usize] = Some(SceneData {
                        index,
                        name,
                    });
                }
                return true;
            }
        }
        false
    }

    pub fn get_state(&self) -> SceneManagerState {
        let scenes = self.scenes.iter().filter_map(|s| s.clone()).collect();
        SceneManagerState {
            current_scene: self.current_scene.clone(),
            scenes,
        }
    }

    pub fn set_active_scene(&mut self, index: u8) {
        self.active_scene_index = index;
        if let Some(ref mut cs) = self.current_scene {
            cs.index = index;
            if (index as usize) < self.scenes.len() {
                if let Some(ref s) = self.scenes[index as usize] {
                    cs.name = s.name.clone();
                }
            }
        }
    }
}
